supportedBackends/spark/src/main/scala/cromwell/backend/impl/spark/SparkClusterProcess.scala

```diff
@@ -0,0 +1,173 @@
+  val RunningStatus = "RUNNING"
+  val FinishedStatus = "FINISHED"
+  val FailedStatus = "FAILED"
+  val FAILED = -1
+  val FINISHED = 0
+  val sleepTime = 1000
+  val submissionIdInJson = "submissionId"
+  val successInJson = "success"
+  val submissionsApiLink = "v1/submissions/status"
+  lazy val SparkClusterMasterHostName = Try(sys.env("HOSTNAME")) match {
+    case Success(s) => Some(s)
+    case Failure(_) => None
+  }
+  val sparkClusterSubmissionResponseRegex = """(?s)\{(.*?)}""".r.unanchored
+  lazy val SparkClusterMasterRestUri = SparkClusterMasterHostName
+    .map(p => String.format("%s%s%s", HttpProtocol, p, s":$SparkRestApiPort"))
+    .getOrElse(String.format("%s%s%s", HttpProtocol, SparkClusterMasterDefault, s":$SparkRestApiPort"))
+}
+
+trait SparkClusterRestClient {
+  def sendAndReceive: SendReceive
+  def makeHttpRequest(httpRequest: HttpRequest): Future[HttpResponse]
+}
+
+trait SparkClusterProcessMonitor {
+  def startMonitoringSparkClusterJob(jobPath: Path,  jsonFile: String): Future[TerminalStatus]
+  def monitorSparkClusterJob(subId: String, rcPath: Path, promise: Promise[Unit]): Unit
+  def evaluateMonitoringFuture(rcPath: Path): Unit
+  def completeMonitoringProcess(rcPath: Path, status: String, promise: Promise[Unit]): Unit
+}
+
+trait SparkClusterJobParser {
+  def parseJsonForSubmissionIdAndStatus(jsonFile: Path): ParserResponse
+}
+
+class SparkClusterProcess(implicit system: ActorSystem) extends SparkProcess
+  with SparkClusterRestClient with SparkClusterJobParser with SparkClusterProcessMonitor {
+
+  import SparkClusterConstants._
+  import SparkClusterProcess._
+  import spray.httpx.SprayJsonSupport._
+  import SparkClusterJsonProtocol._
+
+  implicit lazy val ec: ExecutionContext = system.dispatcher
+  lazy val completionPromise = Promise[TerminalStatus]()
+  lazy val monitorPromise = Promise[Unit]()
+  val tag = this.getClass.getSimpleName
+  lazy val logger = Logger(LoggerFactory.getLogger(getClass.getName))
+
+  override def sendAndReceive: SendReceive = sendReceive
+
+  override def startMonitoringSparkClusterJob(jobPath: Path,  jsonFile: String): Future[TerminalStatus] = {
+    Future(parseJsonForSubmissionIdAndStatus(jobPath.resolve(jsonFile))) onComplete  {
+      case Success(resp: SuccessfulParsedResponse) =>
+        val rcPath = jobPath.resolve(ReturnCodeFile)
+        monitorSparkClusterJob(resp.submissionId, rcPath, monitorPromise)
+        evaluateMonitoringFuture(rcPath)
+      case Failure(exception: Throwable) =>
+        logger.error(s"{} Spark Job failed to submit successfully following reason: {}", tag, exception.getMessage)
+        completionPromise success Failed(exception)
+    }
+    completionPromise.future
+  }
+
+  override def monitorSparkClusterJob(subId: String,  rcPath: Path, promise: Promise[Unit]): Unit = {
+    pollForJobStatus(subId) onComplete {
+      case Success(resp: SuccessResponse) if resp.driverState == RunningStatus => Thread.sleep(sleepTime)
```

Comment:
Thread is owned by the execution actor but since we are using promise in the execution and returning future[backendresponse] from execute, we will complete that promise with this future once job is finished or failed.